File: packages/2021/src/01.ts
import { AoC, toNumberArray } from 'aocf';

export class Aoc2021Day1 extends AoC<number[]> {
  constructor() {
    super(2021, 1);
  }

  parse(input: string): number[] {
    return toNumberArray(input);
  }

  partA(data: number[]): number {
    let count = 0;
    for (let i = 0; i < data.length - 1; i++) {
      if (data[i + 1] > data[i]) count++;
    }
    return count;
  }

  partB(data: number[]): number {
    let count = 0;
    const len = data.length - 3;
    for (let i = 0; i < len; i++) {
      if (data[i + 3] > data[i]) count++;
    }
    return count;
  }
}

export const aoc = new Aoc2021Day1();

aoc.test((o) => {
  const testValues = [199, 200, 208, 210, 200, 207, 240, 269, 260, 263];

  o('PartA', () => {
    o(aoc.partA(testValues)).equals(7);
  });

  o('PartB', () => {
    o(aoc.partB(testValues)).equals(5);
  });
});

File: packages/2021/rust/01.rs
mod aoc;

#[inline]
fn puzzle_a(input: &Vec<usize>) -> usize {
    let mut count: usize = 0;
    for i in 0..input.len() - 1 {
        if input[i + 1] > input[i] {
            count = count + 1
        }
    }
    return count;
}

#[inline]
fn puzzle_b(input: &Vec<usize>) -> usize {
    let mut count = 0;
    for i in 0..input.len() - 3 {
        if input[i + 3] > input[i] {
            count = count + 1
        }
    }
    return count;
}

fn puzzle_parse(input: &String) -> Vec<usize> {
    return input
        .lines()
        .map(|line| line.parse::<usize>().unwrap())
        .collect();
}

fn main() {
    let user_name = "blacha";
    let puzzle = aoc::puzzle_load(&user_name, 2021, 1);
    let data = puzzle_parse(&puzzle.input);

    println!(
        "Puzzle {}-{} a:{} b:{}",
        puzzle.year,
        puzzle.day,
        puzzle.a.unwrap(),
        puzzle.b.unwrap()
    );

    let res_a = puzzle_a(&data);
    if let Some(ans_a) = puzzle.a {
        assert_eq!(res_a, ans_a);
    }

    let res_b = puzzle_b(&data);
    if let Some(ans_b) = puzzle.b {
        assert_eq!(res_b, ans_b);
    }
}
